### PR TITLE
KTOR-572 Retry on HttpCode or network error

### PR DESCRIPTION
**Pull request by cy6erGn0m**
**Subsystem**
ktor-client-retry

**Motivation**
[KTOR-572 Retry on HttpCode or network error](https://youtrack.jetbrains.com/issue/KTOR-572)
[KTOR-945 Apache: ConnectionClosedException when requesting endpoint with the same client two times with some delay between](https://youtrack.jetbrains.com/issue/KTOR-945)

Currently, when the client fails to connect or a pooled connection is dead for some reason, the request fails. Also, sometimes, an error status code could be recovered by retrying the request.

**Solution**
Provide an optional feature that does retry on the particular list of exceptions to recover from I/O errors and request validation errors.

### CODE REVIEW

File: ktor-client/ktor-client-tests/jvm/src/io/ktor/client/tests/utils/ClientLoaderJvm.kt [hunk 1]
```diff
@@ -24,7 +24,7 @@ public actual abstract class ClientLoader {
     public lateinit var engine: HttpClientEngineContainer
 
     @get:Rule
-    public open val timeout: CoroutinesTimeout = CoroutinesTimeout.seconds(60)
+    public open val timeout: CoroutinesTimeout = CoroutinesTimeout.seconds(5)
```
Comment (rsinukov) [inline ktor-client/ktor-client-tests/jvm/src/io/ktor/client/tests/utils/ClientLoaderJvm.kt]: is it intentional?

File: ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Exceptions.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        collectedErrors.forEach { subCause ->
+            cause.addSuppressed(subCause)
+        }
+    }
+
+    throw cause
+}
+
+
+internal fun Result<*>.checkExceptionOrThrowIfNotWhitelisted(): Throwable? {
```
Comment (rsinukov) [inline ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Exceptions.kt:30]: whitelist -> allowed or something similar

File: ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Retry.kt [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+ * The number of retries is limited by [retries] property.
+ * After every failed attempt it does a fixed time delay of [retryIntervalInSeconds].
+ *
+ * When all attempts exceeded, [RequestRetriesExceededException] exception is thrown
+ * having a list of all errors in the suppressed list and/or a cause.
+ *
+ * @property retries number of retries to recover
+ * @property retryIntervalInSeconds between attempts
+ */
+public class Retry private constructor(
```
Comment (rsinukov) [inline ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Retry.kt:34]: A couple of design questions: 
* Should the list of allowed exceptions be configurable?
* Can we enable/disable retries per request? Can be a flag in request builder that saves some flag in the request attributes

**Review by rsinukov — CHANGES_REQUESTED** [review]

File: ktor-client/ktor-client-tests/jvm/src/io/ktor/client/tests/utils/ClientLoaderJvm.kt (hunk 1 above)
Comment (cy6erGn0m) [reply ktor-client/ktor-client-tests/jvm/src/io/ktor/client/tests/utils/ClientLoaderJvm.kt]: Sorry, accidental change, reverted

File: ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Retry.kt (hunk 3 above)
Comment (cy6erGn0m) [reply ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Retry.kt:34]: Don't know how to configure it as people usually need subtypes that is not possible with KClass while lambdas have problems with freezes. Any ideas?
Regarding the flag: any ideas, what use-cases? How it should look like? Only inside of the request builder block, right?

File: ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Exceptions.kt (hunk 2 above)
Comment (cy6erGn0m) [reply ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Exceptions.kt:30]: IfNotRecoverable ?

File: ktor-client/ktor-client-features/ktor-client-retry/api/ktor-client-retry.api [hunk 4]
```diff
@@ -0,0 +1,38 @@
+public final class io/ktor/client/fatures/retry/Retry {
```
Comment (e5l) [inline ktor-client/ktor-client-features/ktor-client-retry/api/ktor-client-retry.api:1]: package name typo

File: ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Exceptions.kt (hunk 2 above)
Comment (e5l) [reply ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Exceptions.kt:30]: I would also avoid extension on `Result` type

File: ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Exceptions.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        is IOException,
+        is HttpRequestTimeoutException,
+        is ResponseException,
+        is UnresolvedAddressException -> cause
+        else -> throw cause
+    }
+}
+
+internal fun Throwable.unwrapCancellation(): Throwable = when (this) {
+    is CancellationException -> cause?.unwrapCancellation() ?: this
```
Comment (e5l) [inline ktor-client/ktor-client-features/ktor-client-retry/common/src/io/ktor/client/fatures/retry/Exceptions.kt:43]: Can we get an infinite loop here?

**Review by e5l — CHANGES_REQUESTED** [review]
I would delay this feature and make the design first

**Comment by cy6erGn0m** [comment]
Closing it for now. Will reopen later with 1.6.0

**Pull request closed without merging**